enforcer/src/main/java/org/wso2/micro/gateway/enforcer/server/WebSocketHandler.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.micro.gateway.enforcer.server;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.wso2.micro.gateway.enforcer.api.APIFactory;
+import org.wso2.micro.gateway.enforcer.api.RequestContext;
+import org.wso2.micro.gateway.enforcer.api.WebSocketAPI;
+import org.wso2.micro.gateway.enforcer.api.WebSocketMetadataContext;
+import org.wso2.micro.gateway.enforcer.api.config.APIConfig;
+import org.wso2.micro.gateway.enforcer.constants.APIConstants;
+import org.wso2.micro.gateway.enforcer.constants.AuthenticationConstants;
+import org.wso2.micro.gateway.enforcer.security.AuthenticationContext;
+import org.wso2.micro.gateway.enforcer.websocket.RateLimitRequest;
+
+/**
+ * WebSocketHandler handles requests coming through websocket metadata service.
+ */
+public class WebSocketHandler implements RequestHandler<RateLimitRequest, AuthenticationContext> {
+    private static final Logger logger = LogManager.getLogger(WebSocketHandler.class);
+    @Override
+    public AuthenticationContext process(RateLimitRequest request) {
+        WebSocketAPI matchedAPI = APIFactory.getInstance().getMatchedAPI(request);
+        if (matchedAPI == null) {
+            return null;
+        } else if (logger.isDebugEnabled()) {
+            APIConfig api = matchedAPI.getAPIConfig();
+            logger.debug("API {}/{} found in the cache", api.getBasePath(), api.getVersion());
+        }
+        RequestContext webSocketMetadata = buildRequestContext(matchedAPI, request);
+
+        return matchedAPI.processMetadata(webSocketMetadata);
+    }
+
+    private RequestContext buildRequestContext(WebSocketAPI api, RateLimitRequest rateLimitRequest) {
+        try {
+            logger.info("contextString:" + rateLimitRequest.getMetadataContext().getFilterMetadataMap().toString());
+        } catch (Exception e) {
+            logger.error(e);
+        }
+        String streamId = rateLimitRequest.getMetadataContext().getFilterMetadataMap()
+                .get(APIConstants.EXT_AUTHZ_METADATA).getFieldsMap().get(APIConstants.WEBSOCKET_STREAM_ID)
```

Comment:
To avoid code repetition L58-L130 (rateLimitRequest.getMetadataContext().getFilterMetadataMap().get(APIConstants.EXT_AUTHZ_METADATA).getFieldsMap()), Let's initialize a common map and directly assign its values in authentication context

Reply:
What I'm suggesting is, first  initialize map as java.util.Map<String, com.google.protobuf.Value> metaData = rateLimitRequest.getMetadataContext().getFilterMetadataMap().get(APIConstants.EXT_AUTHZ_METADATA).getFieldsMap(), then, refer it within authenticationContext variables as follows.
authenticationContext.setApiName(metaData.get(APIConstants.GW_API_NAME_PARAM).getStringValue());
authenticationContext.setApiVersion(metaData.get(APIConstants.GW_VERSION_PARAM).getStringValue());

Reply:
Fixed